service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+		level := queueCollection.Level()
+		c.processingLock.Lock()
+		if shouldProcess, ok := c.shouldProcess[level]; !ok || !shouldProcess {
+			continue
+		}
+		c.shouldProcess[level] = false
+		c.processingLock.Unlock()
+
+		activeQueue := queueCollection.ActiveQueue()
+		if activeQueue == nil {
+			// process for this queue collection has finished
+			// it's possible that new queue will be added to this collection later though,
+			// pollTime will be updated after split/merge
+			continue
+		}
+
+		readLevel := activeQueue.State().ReadLevel()
+		maxReadLevel := minTaskKey(activeQueue.State().MaxLevel(), c.updateMaxReadLevel())
+		domainFilter := activeQueue.State().DomainFilter()
+
+		if !readLevel.Less(maxReadLevel) {
+			// no task need to be processed for now, wait for new task notification
+			// note that if taskID for new task is still less than readLevel, the notification
+			// will just be a no-op and there's no DB requests.
+			continue
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), loadQueueTaskThrottleRetryDelay)
+		if err := c.rateLimiter.Wait(ctx); err != nil {
+			cancel()
+			if level != defaultProcessingQueueLevel {
+				c.setupBackoffTimer(level)
+			} else {
+				c.readyForProcess(level)
+			}
+			continue
+		}
+		cancel()
+
+		taskInfos, more, err := c.readTasks(readLevel, maxReadLevel)
+		if err != nil {
+			c.logger.Error("Processor unable to retrieve tasks", tag.Error(err))
+			c.readyForProcess(level) // re-enqueue the event
+			continue
+		}
+
+		c.Lock()
+		if c.outstandingTaskCount > c.options.MaxPendingTaskSize() {
+			c.logger.Warn("too many outstanding tasks in cross cluster queue.")
+			c.Unlock()
+			return
+		}
+		c.addOutstandingTaskCountLocked(len(taskInfos))
+		c.Unlock()
+
+		tasks := make(map[task.Key]task.Task)
+		newTaskID := readLevel.(transferTaskKey).taskID
+		for _, taskInfo := range taskInfos {
+			if !domainFilter.Filter(taskInfo.GetDomainID()) {
+				continue
+			}
+			task := c.taskInitializer(taskInfo)
+			tasks[newCrossClusterTaskKey(taskInfo.GetTaskID())] = task
+			newTaskID = task.GetTaskID()
+		}
+
+		var newReadLevel task.Key
+		if !more {
+			newReadLevel = maxReadLevel
+		} else {
+			newReadLevel = newCrossClusterTaskKey(newTaskID)
+		}
+
+		queueCollection.AddTasks(tasks, newReadLevel)
+		newActiveQueue := queueCollection.ActiveQueue()
+		if more || (newActiveQueue != nil && newActiveQueue != activeQueue) {
+			// more tasks for the current active queue or the active queue has changed
+			c.readyForProcess(level)
+		}
+	}
+}
+
+func (c *crossClusterQueueProcessorBase) notifyNewTask() {
+	select {
+	case c.notifyCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *crossClusterQueueProcessorBase) validateOutstandingTasks() {
+	tasks := c.getTasks()
+	for _, task := range tasks {
+		if c.isDomainChanged(task) || c.isSourceWorkflowClosed(task) {
+			_, err := c.submitTask(task)
+			if err != nil {
+				// the queue is shutting down
+				return
+			}
+		}
+	}
+}
```

Comment:
- I am thinking this should probably be a method on the CrossClusterTask itself since different task types may have different conditions. For example, signal/child workflow tasks should be executed only when the workflow is open while recordChild or applyParentClosePolicy tasks should be executed only when the workflow is closed.
- Current `isDomainChanged` only checks for the source domain, I believe we also need to check for the target domain.
- Task version check is also needed I think? Can be done as part of the `IsReadyForPoll` method I think.

Reply:
Moved to task interface.